### PR TITLE
Pull request revisions.

### LINKED ISSUE #5254: Implement Pull Requests history
It would be nice to have history of changes in the pull requests as it is in gitlab, for example: if someone pushes commit with `force`, it would be nice to have diff of what was before this push, so, probably, we should store changes for that and notify user on each push event as in gitlab.

### LINKED ISSUE #12604: Versions in Pull requests.
After someone opened a PR from a branch every push that introduces new commits (including the cases where some or all of the previous commits are removed/changed) should generate an entry, a kind of "new version" in the Pull Request's page.

It should be possible to quickly look at this entry generated and find out what has changed between the versions (by taking the user to a diff/compar page)

This is supported in Github, Gitlab, and Gerrit. 

Also on the changes _Files changed_ tab the user should be able to compare any two versions and see what has changed between those versions.
(This is supported in Gitlab and Gerrit but not in Github at the moment, or at least its not clearly reachable from the PR's page).

### PR DESCRIPTION
**Pull request by ghost**
Fixes #12604 
Fixes #5254

### CODE REVIEW

**Comment by zeripath** [comment]
Lint is failing due to a spelling mistake.

_I before E except after C or sounding as Ay as in Weigh..._ and all that jazz...

File: models/pull.go [hunk 1]
```diff
@@ -457,6 +457,12 @@ func newPullRequestAttempt(repo *Repository, pull *Issue, labelIDs []int64, uuid
 		return fmt.Errorf("Commit: %v", err)
 	}
 
+	_, err = x.Get(pr)
```
[the remark this answers is not in the document]
Comment (05storm26) [reply models/pull.go:460]: We need the `pr.ID` for the revision later.

**Comment by alvardes** [comment]
2 things to note: 

1. The owner/repo/compare/commit1...commit2 doesn't work correctly when the two commit hashes are the same content just the commit different. Example: create a PR with two commits, squash and then force push. Now you have two revisions. Comparing the second to the first one should show no additional change. It actually works perfectly on the Files changed pane in the PR itself. However if you do a full revision compare by clicking on compare to previous on the Revisions page (which will take you to the /compare/headhashv1...headhashv2 url) that will show all the files as changed. Its like a `git show` instead of a `git diff`. I looked at the code and I think maybe the compareinfo is bad?
2. The revision creation is different for the creation of the PR and subsequent pushes. For subsequent pushes the revision is created in two steps. 1 is the git ref creation which is done at the post receive hook. And the database entry is created by the PushPullRequest later because only then is the `pusher` available.

**Comment by alvardes** [comment]
Note to self: a separate patchset/difftree comparison page is needed. (to implement use: git patch, git diff-tree, git merge-base, git cherry)

Currently I believe that our compare page doesn't do the kind of shit that github does when comparing branches and listing how many commits behind ahead is the branch. This is good. A simple compare is basically should be a diff without the noise of that kind of stuff and a separate more detailed tree/patches compare page should exists.

**Comment by alvardes** [comment]
Alright I will leave the diff tree feature to an other day. I will open an issue for it.

File: models/pull.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ghost) [reply models/pull.go]: We create a Revision record for each revision in the database table. These entries reference the Pull request so we need the PR-s id in the database for it.

**Comment by alvardes** [comment]
Okay, I think it is good enough. I might create the more comprehensive tree comparsion page later.

**Comment by alvardes** [comment]
Screenshot from 2020-09-10 16-29-34
Screenshot from 2020-09-10 16-29-28
Screenshot from 2020-09-10 16-29-24
Screenshot from 2020-09-10 16-29-16
Screenshot from 2020-09-10 16-29-07

File: models/revision.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+func CreateNewRevision(pr *PullRequest, user *User, commit string) (*Revision, error) {
+	err := pr.LoadBaseRepo()
+	if err != nil {
+		return nil, err
+	}
+
+	repo, err := git.OpenRepository(pr.BaseRepo.RepoPath())
+	if err != nil {
+		return nil, err
+	}
```
Comment (a1012112796) [inline models/revision.go:183]: defer close

File: models/revision.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	err := pr.LoadBaseRepo()
+	if err != nil {
+		return nil, err
+	}
+
+	baseRepo, err := git.OpenRepository(pr.BaseRepo.RepoPath())
+	if err != nil {
+		return nil, err
+	}
+	var refs, error = baseRepo.GetRevisionRefs(pr.Index)
```
Comment (lunny) [inline models/revision.go:96]: defer baseRepo.Close()

**Comment by alvardes** [comment]
The tree compare issue: #12800

File: models/revision.go (hunk 2 above)
Comment (ghost) [reply models/revision.go:183]: Done

File: routers/private/hook.go [hunk 4]
```diff
@@ -525,6 +525,15 @@ func HookPostReceive(ctx *macaron.Context, opts private.HookOptions) {
 					URL:     fmt.Sprintf("%s/compare/%s...%s", baseRepo.HTMLURL(), util.PathEscapeSegments(baseRepo.DefaultBranch), util.PathEscapeSegments(branch)),
 				})
 			} else {
+				gitRepo, err := git.OpenRepository(baseRepo.RepoPath())
+				if err != nil {
+					defer gitRepo.Close()
+				} else {
```
[the remark this answers is not in the document]
Comment (ghost) [reply routers/private/hook.go]: :D 

Fixed

File: models/revision.go (hunk 3 above)
Comment (ghost) [reply models/revision.go:96]: Done

File: models/pull.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zeripath) [reply models/pull.go]: Am I understanding correctly that the purpose of the Get is to retrieve the ID only?

The ID will be set by xorm Insert() on line 452.

If you'd like to check, add a log line just to prove it.

**Comment by a1012112796** [comment]
@techknowlogick Please detach WIP label :)

File: models/revision.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+			if e != nil {
+				return nil, e
+			}
+
+			if has {
+				rev.PrevChangesetHead = prevrev.Commit
+			}
+		}
+		return rev, nil
+	}
```
Comment (a1012112796) [inline models/revision.go]: Not suggest doing these things in get function, suggest adding other functions to do it. Thanks. 
By the way, if user does not exist, should not return error, use ``NewGhostUser()`` is better.

File: routers/repo/pull.go [hunk 6]
```diff
@@ -554,6 +577,42 @@ func ViewPullCommits(ctx *context.Context) {
...
+
+	if ctx.Written() {
+		return
+	}
+
+	var revisions, err = models.GetRevisions(issue.PullRequest)
+	if err != nil || revisions == nil {
+		ctx.NotFound("ViewRevisions", nil)
+		return
+	}
```
Comment (a1012112796) [inline routers/repo/pull.go]: if err is exist, Maybe it's a SerrverError ？

File: services/pull/pull.go [hunk 7]
```diff
@@ -56,15 +56,20 @@ func NewPullRequest(repo *models.Repository, pull *models.Issue, labelIDs []int6
 		return err
 	}
 
-	notification.NotifyNewPullRequest(pr)
```
Comment (a1012112796) [inline services/pull/pull.go]: why delay?

**Comment by a1012112796** [comment]
Suggest add some test code, Thanks.

File: models/pull.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ghost) [reply models/pull.go]: I checked it and you are correct. I have removed this "reloading".

File: models/revision.go (hunk 5 above)
Comment (ghost) [reply models/revision.go]: Refactored. The user/ghostuser is fixed.

File: models/migrations/v151.go [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	"code.gitea.io/gitea/modules/timeutil"
+	"xorm.io/xorm"
+)
+
+func addRevisionTable(x *xorm.Engine) error {
+	type Revision struct {
+		ID              int64              `xorm:"pk autoincr"`
+		Commit          string             `xorm:"NOT NULL"`
```
Comment (lunny) [inline models/migrations/v151.go]: I think we may need start commit id and end commit id because it maybe a force push.

Comment (ghost) [reply models/migrations/v151.go]: Why is the start commit id interesting? That is easly computable form the head commit. We store the number of commits. Resolving "Commit~(NumberOfCommits - 1)" rev in git will give you that.

**Comment by alvardes** [comment]
I am not exactly sure why the tests with my fixtures are not working. @a1012112796 any idea how to fix the current unit test failures?

File: models/migrations/v151.go (hunk 8 above)
Comment (lunny) [reply models/migrations/v151.go]: But once you force push something, you cannot caculate them any more.

Comment (ghost) [reply models/migrations/v151.go]: Yes you can. Why do you think you can't? 

1. If you have loose or no longer used commits they are not deleted immediatly, only at the next run of `git gc`
2. If you have a `ref` to something then that doesn't count as "loose" by `git gc`. That is why we are creating the refs `refs/pull/%d/revision/%d` for each revision head. Oviously commits that are referenced by other commits which you keep must also be kept. So the first commit, or in fact all of the commits of each pull revision is obtainable. Using that you can also always calculate whether the new revision is a force push or not.

Comment (ghost) [reply models/migrations/v151.go]: @lunny ?

**Comment by alvardes** [comment]
I added tests and fixed the issues. 

@lunny  @zeripath  @a1012112796  Anything else?

File: models/migrations/v152.go [hunk 9]
```diff
@@ -0,0 +1,23 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"code.gitea.io/gitea/modules/timeutil"
+	"xorm.io/xorm"
```
Comment (lunny) [inline models/migrations/v152.go]: blank line should between internal packages and external packages.

File: models/revision.go [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+	for _, ref := range refs {
+		var index = git.GetRevisionIndexFromRef(ref.Name)
+
+		if index == nil {
+			continue
+		}
+
+		rev, err := GetRevision(pr, *index)
+
+		if err != nil {
```
Comment (lunny) [inline models/revision.go:112]: We can log the error

File: models/revision.go [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+			continue
+		}
+
+		err = rev.LoadUser()
+
+		if err != nil {
```
Comment (lunny) [inline models/revision.go:119]: As the same.

File: models/revision.go [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+			continue
+		}
+
+		err = rev.LoadPullRequest()
+
+		if err != nil {
```
Comment (lunny) [inline models/revision.go:126]: As the same.

File: models/revision.go [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+	"sort"
+
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/timeutil"
+)
+
+// Revision Tracking information about PR revisions.
+type Revision struct {
+	ID                   int64              `xorm:"pk autoincr"`
+	Commit               string             `xorm:"NOT NULL"`
```
Comment (lunny) [inline models/revision.go:17]: What's the `Commit`'s meaning? Could you add some comments here?

File: models/revision.go (hunk 12 above)
Comment (ghost) [reply models/revision.go:126]: done

File: models/revision.go (hunk 11 above)
Comment (ghost) [reply models/revision.go:119]: done

File: models/revision.go (hunk 10 above)
Comment (ghost) [reply models/revision.go:112]: done

File: models/migrations/v152.go (hunk 9 above)
Comment (ghost) [reply models/migrations/v152.go]: done

**Comment by alvardes** [comment]
I think I have run out of time to make this work

**Pull request closed without merging**